hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/DirectoryDeletingService.java

```diff
@@ -0,0 +1,285 @@
+    this.runCount = new AtomicLong(0);
+    this.pathLimitPerTask = ozoneManager.getConfiguration()
+        .getInt(OZONE_PATH_DELETING_LIMIT_PER_TASK,
+            OZONE_PATH_DELETING_LIMIT_PER_TASK_DEFAULT);
+  }
+
+  private boolean shouldRun() {
+    if (ozoneManager == null) {
+      // OzoneManager can be null for testing
+      return true;
+    }
+    return ozoneManager.isLeaderReady();
+  }
+
+  private boolean isRatisEnabled() {
+    if (ozoneManager == null) {
+      return false;
+    }
+    return ozoneManager.isRatisEnabled();
+  }
+
+  @Override
+  public BackgroundTaskQueue getTasks() {
+    BackgroundTaskQueue queue = new BackgroundTaskQueue();
+    queue.add(new DirectoryDeletingService.DirDeletingTask());
+    return queue;
+  }
+
+  private class DirDeletingTask implements BackgroundTask {
+
+    @Override
+    public int getPriority() {
+      return 0;
+    }
+
+    @Override
+    public BackgroundTaskResult call() throws Exception {
+      if (shouldRun()) {
+        runCount.incrementAndGet();
+        long count = pathLimitPerTask;
+        try {
+          long startTime = Time.monotonicNow();
+          // step-1) Get one pending deleted directory
+          OmKeyInfo pendingDeletedDirInfo = keyManager.getPendingDeletionDir();
+          if (pendingDeletedDirInfo != null) {
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Pending deleted dir name: {}",
+                  pendingDeletedDirInfo.getKeyName());
+            }
+            // step-1: get all sub directories under the deletedDir
+            List<OmKeyInfo> dirs =
+                keyManager.getPendingDeletionSubDirs(pendingDeletedDirInfo,
+                    --count);
+            count = count - dirs.size();
+            List<OmKeyInfo> deletedSubDirList = new ArrayList<>();
+            for (OmKeyInfo dirInfo : dirs) {
+              deletedSubDirList.add(dirInfo);
+              if (LOG.isDebugEnabled()) {
+                LOG.debug("deleted sub dir name: {}",
+                    dirInfo.getKeyName());
+              }
+            }
+
+            // step-2: get all sub files under the deletedDir
+            List<OmKeyInfo> purgeDeletedFiles =
+                keyManager.getPendingDeletionSubFiles(pendingDeletedDirInfo,
+                    --count);
+
+            if (LOG.isDebugEnabled()) {
+              for (OmKeyInfo fileInfo : purgeDeletedFiles) {
+                LOG.debug("deleted sub file name: {}", fileInfo.getKeyName());
+              }
+            }
+
+            // step-3: Since there is a boundary condition of 'numEntries' in
+            // each batch, check whether the sub paths count reached batch size
+            // limit. If count reached limit then there can be some more child
+            // paths to be visited and will keep the parent deleted directory
+            // for one more pass.
+            List<String> purgeDeletedDirs = new ArrayList<>();
+            if (count > 0) {
```

Comment:
Here we at most purge one deleted dir, seems no need to use the List to store this.

Reply:
Thanks @linyiqun for the reviews. Yes, you are perfectly correct. 

**Design Thoughts:** I've plans to enhance the implementations by adding more directories by recursively traversing to it sub-paths if there are less number of sub-path components in the tree. Please find the below TODO comment in the patch. I'm pasting the same thing here as well for our reference.

```
            if (count > 0) {
              // TODO: Now, there is only one entry in this list. Maintained
              //  list data structure becuase this can be extended to add
              //  more directories within the batchSize limit.
              purgeDeletedDirs.add(pendingDeletedDirInfo.getPath());
            }
```

Assume below is the FS dir structure at OM and batch size is 1000. Again, assume user has deleted dir 'a'. 
With recursive listing in place, `DirDeletingService` can collect dirs [a, b1, b2, b3, c1, c2, d1, d2, d3, e1, e2, e3] in first pass then delete these from `DirectoryTable` and files [d21.txt, d22.txt, e11.txt] will be added to `DeletedTable` so that KDS can take care its block(s) deletions. 
This way I can avoid adding the sub-directory child entries into `DeletedDirectoryTable` and save IO(OM Database) operations. Since this involves some complexity I am planning to implement it via another jira task. Probably, this along with other performance optimizations can be done after feature branch merge to master. Let's see how quickly we can complete the major functionality parts and reach to a logical completion of the feature:-)

By keeping all these thoughts in mind I've used list data structure:-)
Does this make sense to you and shall I keep list data structure?

```
     *                  buck-1
     *                    |
     *                    a
     *                    |
     *      -----------------------------------
     *     |           |                       |
     *     b1          b2                      b3
     *   -----       --------               ----------
     *   |    |      |    |   |             |    |     |
     *  c1   c2     d1   d2  d3             e1   e2   e3
     *                   |                  |
     *               --------               |
     *              |        |              |
     *           d21.txt   d22.txt        e11.txt
```

Reply:
FYI, I've created task HDDS-5048 to revisit this part later

Reply:
Agreed to revisit this in another task.

>With recursive listing in place, DirDeletingService can collect dirs [a, b1, b2, b3, c1, c2, d1, d2, d3, e1, e2, e3]..

Current problem is that we delete dir from parent dir to child, it's not efficient because parent dir relied on subdir to be purged firstly. So then parent dir won't hit the count limitation. But this can lead the repeated traversing on parent dir.  A better way I am thinking we could use recursive way to collect all dirs and then stored in a stack structure. 
For above example, the dirs will be stored as[a, b1, c1, c2, b2, d1, d2, d3, b3, e1, e2, e3]. Then we pop the dir from the stack, the delete order is from e3, e2, ..., a. 

Based on above way, we could make sure there is no any subdir needed to be purged when purging a dir.

Reply:
@linyiqun Thanks for adding more details.

>A better way I am thinking we could use recursive way to collect all dirs and then stored in a stack structure.
Say, if user has deleted a very large parent dir with millions of sub-children with more depth(say 10 levels) and each level has more than batchSized dirs and files. I think, maintaining stack will have complexity by increasing heap memory consumption.

> Current problem is that we delete dir from parent dir to child, it's not efficient because parent dir relied on subdir to be purged firstly
I am deleting the parentdir once all its immediate sub-dirs are added to `DeletedDirectortTable` and sub-files to `DeletedTable`. This ensures that parentdir can be safely removed and the sub-dirs will become/promotted as parentdirs for the next DDService iteration.

Appreciate your feedback. Thanks again!

Reply:
@linyiqun Does my above comment make sense to you. Thanks!

Reply:
Comment makes the sense to me, @rakeshadr .